cpp/include/raft/distance/distance.cuh

```diff
@@ -0,0 +1,310 @@
+           Index_ k, void *workspace, size_t worksize, FinalLambda fin_op,
+           cudaStream_t stream, bool isRowMajor) {
+    raft::distance::euclideanAlgo2<InType, AccType, OutType, FinalLambda,
+                                   Index_>(m, n, k, x, y, dist, false, fin_op,
+                                           stream, isRowMajor);
+  }
+};
+
+template <typename InType, typename AccType, typename OutType,
+          typename FinalLambda, typename Index_>
+struct DistanceImpl<raft::distance::DistanceType::L2SqrtUnexpanded, InType,
+                    AccType, OutType, FinalLambda, Index_> {
+  void run(const InType *x, const InType *y, OutType *dist, Index_ m, Index_ n,
+           Index_ k, void *workspace, size_t worksize, FinalLambda fin_op,
+           cudaStream_t stream, bool isRowMajor) {
+    raft::distance::euclideanAlgo2<InType, AccType, OutType, FinalLambda,
+                                   Index_>(m, n, k, x, y, dist, true, fin_op,
+                                           stream, isRowMajor);
+  }
+};
+
+template <typename InType, typename AccType, typename OutType,
+          typename FinalLambda, typename Index_>
+struct DistanceImpl<raft::distance::DistanceType::L1, InType, AccType, OutType,
+                    FinalLambda, Index_> {
+  void run(const InType *x, const InType *y, OutType *dist, Index_ m, Index_ n,
+           Index_ k, void *workspace, size_t worksize, FinalLambda fin_op,
+           cudaStream_t stream, bool isRowMajor) {
+    raft::distance::l1Impl<InType, AccType, OutType, FinalLambda, Index_>(
+      m, n, k, x, y, dist, fin_op, stream, isRowMajor);
+  }
+};
+
+}  // anonymous namespace
+
+/**
+ * @brief Return the exact workspace size to compute the distance
+ * @tparam DistanceType which distance to evaluate
+ * @tparam InType input argument type
+ * @tparam AccType accumulation type
+ * @tparam OutType output type
+ * @tparam Index_ Index type
+ * @param x first set of points
+ * @param y second set of points
+ * @param m number of points in x
+ * @param n number of points in y
+ * @param k dimensionality
+ *
+ * @note If the specifed distanceType doesn't need the workspace at all, it
+ * returns 0.
+ */
+template <raft::distance::DistanceType distanceType, typename InType,
+          typename AccType, typename OutType, typename Index_ = int>
+size_t getWorkspaceSize(const InType *x, const InType *y, Index_ m, Index_ n,
+                        Index_ k) {
+  size_t worksize = 0;
+  constexpr bool is_allocated =
+    distanceType <= raft::distance::DistanceType::CosineExpanded;
+  if (is_allocated) {
+    worksize += m * sizeof(AccType);
+    if (x != y) worksize += n * sizeof(AccType);
+  }
+  return worksize;
+}
+
+#if CUDART_VERSION >= 10010
+// With optimization enabled, CUDA 10.1 generates segfaults for distance
+// prims, so disable optimization until another workaround is found
+// #pragma GCC push_options
+#pragma GCC optimize("O0")
+#endif
```

Comment:
IIRC, this was added because there were issues with cutlass. Now that we have rid ourselves of it, it should be ok to remove this